### PR TITLE
fix(cdk/a11y): FocusMonitor emits rogue values in IE 11

### LINKED ISSUE #15517: CDK FocusMonitor emits invalid state in IE11
#### What is the expected behavior?
In all other browsers when i monitor an element via:
```typescript
  ngAfterViewInit() {
    this._focusMonitor
      .monitor(this.elementRef, true)
//      .pipe(debounceTime(0))
      .subscribe(focusOrigin => {
        if (!focusOrigin && this.radioGroup) {
          this.radioGroup._touch();
        }
      });
  }
```

And i click on the element, 2 values get emitted with `focusOrigin === 'mouse'`.

The workaround i've found is to add a `debounceTime(0)` like the commented line above.

The code in question is from `radio.ts` in material.

#### What is the current behavior?
On IE11, between the `'mouse'` values, i get a `null` value emitted.

#### What are the steps to reproduce?
Try running the example https://stackblitz.com/angular/kgnogyvppxp?file=app%2Fform-field-label-example.ts on IE11 and a console.log in `radio.ts` on the `focusOrigin`.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Latest material, IE11.

### PR DESCRIPTION
**Pull request by davidjgoss**
Fixes #15517.

Not ready to merge just yet, but I wanted to get this up and get any feedback in case I haven't fully understood the potential impact.

- [x] Fully update documentation
- [x] Rebase and squash

TLDR of the issue:

> When you monitor a parent element with `checkChildren` set to `true`, internally the focus monitor listens for both `focus` and `blur` events on the parent element. When focus shifts from child element 1 to child element 2, the `blur` event on child element 1 is captured first. The design is that this `blur` event does not result in a `null` value on the subject, because the focus monitor will see that the [event's `relatedTarget`](https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent/relatedTarget) (the element that's about to take the focus) is also a child of our parent element, [and take no further action](https://github.com/angular/components/blob/2d976a92429ed37a89bfa9ee74b479b48c7ac40d/src/cdk/a11y/focus-monitor/focus-monitor.ts#L375).
> 
> However, IE does not set the `relatedTarget` on `focus` or `blur` events, so the focus monitor can't know that focus is not leaving the parent element, and so we get a `null` value on the subject.

The solution here is to listen for the `focusin` and `focusout` events instead, where IE _does_ set the `relatedTarget` correctly (along with all other browsers). Other than this, there are some differences between the two sets of events:

- `focusin` and `focusout` support bubbling, where `focus` and `blur` do not - I don't think this makes any difference for our use case, as we are explicitly using capture mode anyway
- `focusin` and `focusout` fire _after_ their counterparts `focus` and `blur` - it's theoretically possible the some users may be relying on an order of events, but given there's no documentation that spells out when the focus monitor emits in relation to, say, model changes, I don't think it should be considered a breaking change

(I've not been able to run the full test suite on Windows due to #16632, and linting on Windows is a non-starter because of line breaks, so I'll be testing properly on my Mac later)

### CODE REVIEW

**Comment by jelbourn** [comment]
If you're able to get this PR green, I can presubmit it through Google to see if ti breaks anything there

**Comment by davidjgoss** [comment]
@jelbourn great thanks, will work on that now

**Comment by davidjgoss** [comment]
@googlebot I fixed it.

**Comment by jelbourn** [comment]
Looks like there's still one unit test failing:
https://circleci.com/gh/angular/components/76751?

**Comment by davidjgoss** [comment]
Sorry, getting back onto this now. With the test in Firefox I can observe the correct behaviour in the demo app, so perhaps we aren't firing quite the right combination of events in the test, I'll keep digging.

File: src/cdk/a11y/focus-monitor/focus-monitor.ts [hunk 1]
```diff
@@ -185,14 +185,18 @@ export class FocusMonitor implements OnDestroy {
     let focusListener = (event: FocusEvent) => this._onFocus(event, nativeElement);
     let blurListener = (event: FocusEvent) => this._onBlur(event, nativeElement);
     this._ngZone.runOutsideAngular(() => {
-      nativeElement.addEventListener('focus', focusListener, true);
-      nativeElement.addEventListener('blur', blurListener, true);
+      // @ts-ignore
```
Comment (jelbourn) [inline src/cdk/a11y/focus-monitor/focus-monitor.ts]: Why are these `@ts-ignore` comments here?

Comment (mmalerba) [reply src/cdk/a11y/focus-monitor/focus-monitor.ts]: I suspect because they type is as a union `'focus' | 'blur' | 'click' | ...`, cast `'focusin' as any` is probably better than `@ts-ignore`

Comment (jelbourn) [reply src/cdk/a11y/focus-monitor/focus-monitor.ts]: `focusin` is a totally valid event for `addEventListener` (which falls back to `string` if for custom events)

Comment (davidjgoss) [reply src/cdk/a11y/focus-monitor/focus-monitor.ts]: This is about the signature of the handler rather than the event type argument.

The `focusListener` and `blurListener` functions both take a `FocusEvent` as their argument - they have to, because they need to get at the `relatedTarget` property in order for this functionality to work (understanding where focus is going from _and_ to).

The typings in the version of typescript we are using call out the `focus` and `blur` events as producing a `FocusEvent`, but not `focusin` and `focusout` so they fall back to `Event`. This was incorrect ([per MDN](https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent), all 4 events produce a `FocusEvent`) and has [since been fixed](https://github.com/microsoft/TypeScript/blame/e48cd3a1011ba0b61edee0cc19201cff8ff6f6e6/lib/lib.dom.d.ts#L5810), so the `@ts-ignore` is a temporary workaround.

Comment (jelbourn) [reply src/cdk/a11y/focus-monitor/focus-monitor.ts]: Ack, I would prefer to use `as any` with a TODO comment explaining this rather than `@ts-ignore` since the latter is a bigger hammer

Comment (davidjgoss) [reply src/cdk/a11y/focus-monitor/focus-monitor.ts]: Okay I’ll make that change

Comment (davidjgoss) [reply src/cdk/a11y/focus-monitor/focus-monitor.ts]: Reworked that bit and managed to avoid both `@ts-ignore` and `any`

**Comment by davidjgoss** [comment]
Thanks, I'll clean up the commits

**Comment by davidjgoss** [comment]
@jelbourn do you need me to do anything else in order to get this merge-safe?

**Comment by mmalerba** [comment]
`merge-safe` is a label we use to indicate that a PR is guaranteed not to affect test inside Google. It doesn't apply in this case because this PR will affect tests in Google. You don't need to do anything else for now. We'll run the tests in Google and merge it if they pass or ask for changes if the PR needs to be updated

**Comment by davidjgoss** [comment]
Hey folks. Given that:

- This is fix for an IE-specific bug
- There's a straightforward workaround described in #15517
- Angular's IE support is coming to an end later this year per https://github.com/angular/angular/issues/41840

Should this just be closed?

**Comment by jelbourn** [comment]
Yeah- looking into it, this PR stalled out because it caused a number of tests to start failing inside Google and nobody ever found the time to debug those failures. We certainly could have communicated that better. Closing for now since IE11 isn't long for this world.
